nengo/linear_system.py

```diff
@@ -0,0 +1,369 @@
+import numpy as np
+
+from nengo.base import Process
+from nengo.exceptions import ValidationError
+from nengo.params import BoolParam, EnumParam, NdarrayParam, Parameter
+from nengo.rc import rc
+from nengo._vendor.scipy.signal import (
+    cont2discrete,
+    ss2tf,
+    tf2ss,
+    tf2zpk,
+    zpk2ss,
+)
+
+
+def _as2d(M, last=True, dtype=np.float64):
+    if M is None:
+        return M
+
+    M = np.asarray(M, dtype=dtype)
+    if M.ndim < 2:
+        return M.reshape((1, -1)) if last else M.reshape((-1, 1))
+    else:
+        return M
+
+
+class LinearSystem(Process):
+    """General linear time-invariant (LTI) system.
+
+    Parameters
+    ----------
+    sys : list or tuple
+        A list or tuple representing a system in one of the following forms:
+
+            * length 2: (numerator, denominator) transfer function form [1]_
+            * length 3: (zero, pole, gain) zero-pole-gain form [2]_
+            * length 4: (A, B, C, D) state-space form [3]_
+
+        For more details on these forms, see `scipy.signal.lti`.
+
+    x0 : array_like
+        Initial values for the system state. The last dimension must equal the
+        ``state_size``.
+    analog : boolean, optional
+        Whether ``sys`` is in analog (continuous-time) or discrete form.
+    method : string
+        The method to use for discretization (if ``analog`` is True). See
+        `scipy.signal.cont2discrete` for information about the options.
+
+    Attributes
+    ----------
+    analog : boolean
+        Whether the system is described in analog (continuous-time) or discrete form.
+        Analog coefficients will be converted to discrete using the timestep ``dt``.
+    method : string
+        The method to use for discretization (if ``analog`` is True). See
+        `scipy.signal.cont2discrete` for information about the options.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Transfer_function
+    .. [2] https://en.wikipedia.org/wiki/Pole%E2%80%93zero_plot
+    .. [3] https://en.wikipedia.org/wiki/State-space_representation
+    """
+
+    A = NdarrayParam("A", shape=("*", "*"))
+    B = NdarrayParam("B", shape=("*", "*"))
+    C = NdarrayParam("C", shape=("*", "*"))
+    D = NdarrayParam("D", shape=("*", "*"))
+    x0 = NdarrayParam("x0", shape=("...",))
+    analog = BoolParam("analog")
+    method = EnumParam(
+        "method", values=("bilinear", "euler", "backward_diff", "zoh", "impulse")
+    )
+
+    @classmethod
+    def _find_size(cls, matrices, index):
+        for M in matrices:
+            if M is not None and M.size > 0:
+                return M.shape[index]
+
+        return 0
+
+    def __init__(self, sys, analog=True, method="zoh", x0=0, default_dt=0.001):
+        self._CombineClass = LinearSystem
```

Comment:
Since this is used in both combine and discretize, calling it `_CombineClass` is a bit confusing. Not sure of a better name though.

Reply:
I can't think of a better one ATM either. If you do, let me know, otherwise I'll keep it as-is for now.

Reply:
I think `_factory` might be a bit more telling of its purpose, even if that might not be the most accurate use of the 'factory method pattern' terminology (since it's not really a method that's creates instances of a class). But it functions in that sort of way. I'll add a commit for this and if you or anyone else doesn't like it feel free to revert. :)